android/PhysicalWeb/app/src/main/java/org/physical_web/physicalweb/NearbyBeaconsFragment.java

```diff
@@ -363,34 +371,55 @@ public void onAnimationCancel(Animator animation) {
    */
   private void safeNotifyChange() {
     if (getListView().getVisibility() == View.VISIBLE) {
-      mNearbyDeviceAdapter.notifyDataSetChanged();
+      new Handler(Looper.getMainLooper()).post(new Runnable() {
+        @Override
+        public void run() {
+          mNearbyDeviceAdapter.notifyDataSetChanged();
+        }
+      });
     }
   }
 
   // Adapter for holding beacons found through scanning.
   private class NearbyBeaconsAdapter extends BaseAdapter {
-    private BeaconDisplayList mBeaconDisplayList;
+    private List<PwPair> mPwPairs;
 
     NearbyBeaconsAdapter() {
-      mBeaconDisplayList = new BeaconDisplayList();
+      mPwPairs = new ArrayList<>();
     }
 
-    public void addItem(PwoMetadata pwoMetadata) {
-      mBeaconDisplayList.addItem(pwoMetadata);
+    public void addItem(PwPair pwPair) {
+      mPwPairs.add(pwPair);
     }
 
-    public boolean containsUrl(String url) {
-      return mBeaconDisplayList.containsUrl(url);
+    public void updateItem(PwPair pwPair) {
```

Comment:
Would it make sense to combine addItem and updateItem into one method that does the right thing regardless of whether the item is already in the collection?